sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java

```diff
@@ -171,8 +122,64 @@ private void setNotNullAt(int i) {
   }
 
   @Override
-  public void update(int ordinal, Object value) {
-    throw new UnsupportedOperationException();
+  public void update(int i, Object value) {
+    if (value == null) {
+      if (!isNullAt(i)) {
+        // remove the old value from pool
+        long idx = getLong(i);
+        if (idx <= 0) {
+          // this is the index of old value in pool, remove it
+          pool.replace((int)-idx, null);
+        } else {
+          // there will be some garbage left (UTF8String or byte[])
```

Comment:
We only compare UnsafeRow as bytes in key ofBytesToBytesMap, the key cannot be changed in this case. Otherwise, we will use the general way to compare UnsafeRow with other Rows (calling InternalRow.equals() and hashCode()).